### PR TITLE
Replace version with reader cache key in IndicesRequestCache

### LINKED ISSUE #27650: [CI] RecoveryIT.testRecoveryWithConcurrentIndexing fails on master expecting more documents
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/931/console

```
FAILURE 2.01s | RecoveryIT.testRecoveryWithConcurrentIndexing <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: 
   > Expected: <60>
   >      but: was <110>
   > 	at __randomizedtesting.SeedInfo.seed([65E63137F3BA894E:E5B7F564711E5784]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.upgrades.RecoveryIT.assertCount(RecoveryIT.java:204)
   > 	at org.elasticsearch.upgrades.RecoveryIT.testRecoveryWithConcurrentIndexing(RecoveryIT.java:192)
   > 	at java.lang.Thread.run(Thread.java:748)
```

Expecting that the first node have 60 documents, instead has 110:

```java
            case UPGRADED:
                updateIndexSetting(index, Settings.builder().put(INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.getKey(), (String)null));
                asyncIndexDocs(index, 60, 50).get();
                ensureGreen(index);
                assertOK(client().performRequest("POST", index + "/_refresh"));
                assertCount(index, "_only_nodes:" + nodes.get(0), 60); // <-- failure here
                assertCount(index, "_only_nodes:" + nodes.get(1), 60);
                break;
```

Reproduces for me with:

```
gradle :qa:rolling-upgrade:v6.2.0-SNAPSHOT#upgradedClusterTestRunner -Dtests.seed=65E63137F3BA894E -Dtests.class=org.elasticsearch.upgrades.RecoveryIT -Dtests.method="testRecoveryWithConcurrentIndexing" -Dtests.security.manager=true -Dtests.locale=sl -Dtests.timezone=SST
```

### PR DESCRIPTION
**Pull request by dnhatn**
Today we use the version of a DirectoryReader as a component of the key
of IndicesRequestCache. This usage is perfectly fine since the version
is advanced every time a new change is made into IndexWriter. In other
words, two DirectoryReaders with the same version should have the same
content. However, this invariant is only guaranteed in the context of a
single IndexWriter because the version is reset to the committed version
value when IndexWriter is re-opened.

Since #33473, each IndexShard may have more than one IndexWriter, and
using the version of a DirectoryReader as a part of the cache key can
cause IndicesRequestCache to return stale cached values. For example, in
#27650, we rollback the engine (i.e., re-open IndexWriter), index new
documents, refresh, then make a count request, but the search layer
mistakenly returns the count of the DirectoryReader of the previous
IndexWriter because the current DirectoryReader has the same version of
the old DirectoryReader even their documents are different. This is
possible because these two readers come from different IndexWriters.

This commit replaces the the version with the reader cache key of
IndexReader as a component of the cache key of IndicesRequestCache.

Closes #27650
Relates #33473

/cc @ywelsch

### CODE REVIEW

**Comment by dnhatn** [comment]
Chatted with Boaz on another channel, I'll work on these follow-ups:

1. Index different number of documents in testRecoveryWithConcurrentIndexing
2. Make sure the global checkpoint is always initialized  (never be unassigned)

File: server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java [hunk 1]
```diff
@@ -137,7 +139,9 @@ BytesReference getOrCompute(CacheEntity cacheEntity, Supplier<BytesReference> lo
      * @param cacheKey the cache key to invalidate
      */
     void invalidate(CacheEntity cacheEntity, DirectoryReader reader, BytesReference cacheKey) {
-        cache.invalidate(new Key(cacheEntity, reader.getVersion(), cacheKey));
+        if (reader.getReaderCacheHelper() != null) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java]: Why is it possible to have a null cache helper here ?

**Review by jimczi — COMMENTED** [review]
The changes look good to me. Though I wonder why we were using `getVersion` until now. I don't have the full history so maybe @s1monw can explain if this was needed for something non-obvious. 
One reason I can think of is that the reader cache key changes every time you open a new directory reader even if nothing changed in the index but I don't see where this could be an issue since we reopen the directory reader only if something has changed.

File: server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java (hunk 1 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java]: some reader wrappers will return null here if they change the live-docs of a reader for instance. But I wonder in what case that happens?

**Review by s1monw — APPROVED** [review]
change LGTM. I think the version was really just legacy. I am not sure why we used this. I also can't think of a reason why this change would cause issues.

File: server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/IndicesRequestCache.java]: I've replaced this check with an assertion as we should always have readerCacheKey.

**Comment by dnhatn** [comment]
Thanks @jimczi and @s1monw for reviewing.

**Pull request merged**